components/omega/test/base/HaloTest.cpp

```diff
@@ -0,0 +1,306 @@
+   haloExchangeTest(MyHalo, Init2DR4, Test2DR4, "2DR4", TotErr);
+   haloExchangeTest(MyHalo, Init2DR8, Test2DR8, "2DR8", TotErr);
+
+   // Initialize and run 3D tests
+   for (int K = 0; K < N3; ++K) {
+      for (int ICell = 0; ICell < NCellsAll; ++ICell) {
+         for (int J = 0; J < N2; ++J) {
+            OMEGA::I4 NewVal = (K + 1)*(J + 1)*DefDecomp->CellIDH(ICell);
+            Init3DI4(K, ICell, J) = NewVal;
+            Init3DI8(K, ICell, J) = static_cast<OMEGA::I8>(NewVal);
+            Init3DR4(K, ICell, J) = static_cast<OMEGA::R4>(NewVal);
+            Init3DR8(K, ICell, J) = static_cast<OMEGA::R8>(NewVal);
+         }
+      }
+   }
+
+   haloExchangeTest(MyHalo, Init3DI4, Test3DI4, "3DI4", TotErr);
+   haloExchangeTest(MyHalo, Init3DI8, Test3DI8, "3DI8", TotErr);
+   haloExchangeTest(MyHalo, Init3DR4, Test3DR4, "3DR4", TotErr);
+   haloExchangeTest(MyHalo, Init3DR8, Test3DR8, "3DR8", TotErr);
+
+   // Initialize and run 4D tests
+   for (int L = 0; L < N4; ++L) {
+      for (int K = 0; K < N3; ++K) {
+         for (int ICell = 0; ICell < NCellsAll; ++ICell) {
+            for (int J = 0; J < N2; ++J) {
+               OMEGA::I4 NewVal =
+                  (L + 1)*(K + 1)*(J + 1)*DefDecomp->CellIDH(ICell);
+               Init4DI4(L, K, ICell, J) = NewVal;
+               Init4DI8(L, K, ICell, J) = static_cast<OMEGA::I8>(NewVal);
+               Init4DR4(L, K, ICell, J) = static_cast<OMEGA::R4>(NewVal);
+               Init4DR8(L, K, ICell, J) = static_cast<OMEGA::R8>(NewVal);
+            }
+         }
+      }
+   }
+
+   haloExchangeTest(MyHalo, Init4DI4, Test4DI4, "4DI4", TotErr);
+   haloExchangeTest(MyHalo, Init4DI8, Test4DI8, "4DI8", TotErr);
+   haloExchangeTest(MyHalo, Init4DR4, Test4DR4, "4DR4", TotErr);
+   haloExchangeTest(MyHalo, Init4DR8, Test4DR8, "4DR8", TotErr);
+
+   // Initialize and run 5D tests
+   for (int M = 0; M < N5; ++M) {
+      for (int L = 0; L < N4; ++L) {
+         for (int K = 0; K < N3; ++K) {
+            for (int ICell = 0; ICell < NCellsAll; ++ICell) {
+               for (int J = 0; J < N2; ++J) {
+                  OMEGA::I4 NewVal =
+                     (M + 1)*(L + 1)*(K + 1)*(J + 1)*DefDecomp->CellIDH(ICell);
+                  Init5DI4(M ,L, K, ICell, J) = NewVal;
+                  Init5DI8(M ,L, K, ICell, J) = static_cast<OMEGA::I8>(NewVal);
+                  Init5DR4(M ,L, K, ICell, J) = static_cast<OMEGA::R4>(NewVal);
+                  Init5DR8(M ,L, K, ICell, J) = static_cast<OMEGA::R8>(NewVal);
+               }
+            }
+         }
+      }
+   }
+
+   haloExchangeTest(MyHalo, Init5DI4, Test5DI4, "5DI4", TotErr);
+   haloExchangeTest(MyHalo, Init5DI8, Test5DI8, "5DI8", TotErr);
+   haloExchangeTest(MyHalo, Init5DR4, Test5DR4, "5DR4", TotErr);
+   haloExchangeTest(MyHalo, Init5DR8, Test5DR8, "5DR8", TotErr);
+
+   if (TotErr == 0) {
+      LOG_INFO("HaloTest: Successful completion");
+   } else {
+      LOG_INFO("HaloTest: Failed");
+      return -1;
```

Comment:
This was causing the hang - exiting before MPI_Finalize, so I'd just remove the return statement here.

Reply:
Related to this - I'm getting random FAIL on some of the updates. Still digging.